pyccel/ast/operators.py

```diff
@@ -481,8 +474,38 @@ def __new__(cls, arg1, arg2):
         else:
             return super().__new__(cls)
 
-    def _handle_str_type(self, strs):
-        self._dtype = NativeString()
+    def _calculate_dtype(*_args):
+        """ Sets the dtype and precision
+        If one argument is a string then all arguments must be strings
+        If the arguments are numeric then the dtype and precision
+        match the broadest type and the largest precision
+        e.g.
+            1 + 2j -> PyccelAdd(LiteralInteger, LiteralComplex) -> complex
+        """
+        integers  = [a for a in _args[1] if a.dtype in (NativeInteger(),NativeBool())]
+        reals     = [a for a in _args[1] if a.dtype is NativeReal()]
+        complexes = [a for a in _args[1] if a.dtype is NativeComplex()]
+        strs      = [a for a in _args[1] if a.dtype is NativeString()]
```

Comment:
There is a lot of code duplication here. I think these should be class methods rather than static methods. See [this](https://www.geeksforgeeks.org/class-method-vs-static-method-python/) for a description of static vs class methods. I think with a class method, you should be able to do `cls._handle_integers()` to call the correct version of the method

Reply:
done in [6efbe03](https://github.com/pyccel/pyccel/pull/797/commits/6efbe034730311dd0100d5dd5d9f6444449e94f1)